src/ConnectedMode.UnitTests/ServerEventFilteredChannelTests.cs

```diff
@@ -0,0 +1,152 @@
+{
+    // since TaintServerEventChannel and IssueChangedServerEventChannel only differ in the value of the type parameter and in mef export contract, the tests results are the same for both
+    [TestClass]
+    public class ServerEventFilteredChannelTests
+    {
+        [TestMethod]
+        public void TaintAndIssuesMefCtor_CheckIsExported()
+        {
+            MefTestHelpers.CheckTypeCanBeImported<TaintServerEventChannel, ITaintServerEventSource>();
+            MefTestHelpers.CheckTypeCanBeImported<TaintServerEventChannel, ITaintServerEventSourcePublisher>();
+            MefTestHelpers.CheckTypeCanBeImported<IssueChangedServerEventChannel, IIssueChangedServerEventSource>();
+            MefTestHelpers.CheckTypeCanBeImported<IssueChangedServerEventChannel, IIssueChangedServerEventSourcePublisher>();
+        }
+
+        [TestMethod]
+        public async Task Get_AfterPublish_ReturnsCorrectValue()
+        {
+            var testSubject = CreateTestSubject();
+            var serverEvent = Mock.Of<ITaintServerEvent>();
+
+            testSubject.Publish(serverEvent);
+            var receivedEvent = await testSubject.GetNextEventOrNullAsync();
+
+            receivedEvent.Should().BeSameAs(serverEvent);
+        }
+
+        [TestMethod]
+        public async Task Get_AwaitsForPublishBeforeReturning()
+        {
+            var testSubject = CreateTestSubject();
+            var serverEvent = Mock.Of<ITaintServerEvent>();
+
+            var getTask =  testSubject.GetNextEventOrNullAsync();
+            testSubject.Publish(serverEvent);
+            var receivedEvent = await getTask;
+
+            receivedEvent.Should().BeSameAs(serverEvent);
+        }
+
+        [TestMethod]
+        public async Task Get_ReturnsMultiplePublishedItemsInCorrectOrder()
+        {
+            var testSubject = CreateTestSubject();
+            var serverEvents = Enumerable.Range(0, 5).Select(_ => Mock.Of<ITaintServerEvent>()).ToList();
+            var receivedEvents = new List<ITaintServerEvent>();
+
+            foreach (var serverEvent in serverEvents)
+            {
+                testSubject.Publish(serverEvent);
+            }
+            for (var i = 0; i < serverEvents.Count; i++)
+            {
+                receivedEvents.Add(await testSubject.GetNextEventOrNullAsync());
+            }
+
+            receivedEvents.Should().BeEquivalentTo(serverEvents);
+        }
+
+        [TestMethod]
+        public async Task Get_AlreadyDisposedAndNoMoreItems_ReturnsNull()
+        {
+            var testSubject = CreateTestSubject();
+            var serverEvent = Mock.Of<ITaintServerEvent>();
+
+            testSubject.Publish(serverEvent);
+            testSubject.Dispose();
+            var receivedEvent = await testSubject.GetNextEventOrNullAsync();
+            var receivedEvent2 = await testSubject.GetNextEventOrNullAsync();
+
+            receivedEvent.Should().BeSameAs(serverEvent);
```

Comment:
Ideally, that's _not_ what we would want. If the channel was disposed, any events that are in it are worthless. But I don't think we can clear the channel or anything. Maybe before returning we can check if it's disposed, and if so, not return the item. WDYT? not sure about the threading implications.

Reply:
The bool value is atomic so shouldn't be a problem since we don't need to stop exactly after dispose, but the sooner the better. Btw, this one is a bit unclear for me, should I make it volatile? I always struggled with that keyword 🤔